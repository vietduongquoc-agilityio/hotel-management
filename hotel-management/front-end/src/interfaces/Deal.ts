export interface DealData {
  dealName: string;
  referenceNumber: string;
  startDate: Date;
  endDate: Date;
  roomType: string;
  statusDeal: string;
  reservationsLeft: number;
  documentId: string;
}

export type NewDealData = Omit<DealData, "documentId">;
