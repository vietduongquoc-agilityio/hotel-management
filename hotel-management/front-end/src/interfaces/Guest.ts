export interface GuestData {
    guestName: string;
    roomType: string;
    stay: number;
    price: number;
    registrationNumber: string;
    totalAmount: number;
    checkInDate: Date;
    documentId: string;
}

export type NewGuestData = Omit<GuestData, "documentId">;
