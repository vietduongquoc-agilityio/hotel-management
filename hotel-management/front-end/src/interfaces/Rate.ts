export interface RateData {
  roomType: string;
  cancellationPolicy: string;
  deals: string;
  dealPrice: string;
  availability: string;
  documentId: string;
  rate: string;
  totalOfBooked?: number;
}

export type NewRateData = Omit<RateData, "documentId">;
