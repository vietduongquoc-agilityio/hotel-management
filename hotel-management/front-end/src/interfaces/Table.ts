export interface EditFunctionType<T> {
  initialData: T;
  onEditFunction: (updatedData: T) => void;
}

export interface DeleteFunctionType {
  documentId: string;
  onDeleteFunction: (id: string) => void;
}
