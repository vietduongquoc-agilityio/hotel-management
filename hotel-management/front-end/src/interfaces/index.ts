export * from "./Rate";
export * from "./Room";
export * from "./Table";
export * from "./Guest"
