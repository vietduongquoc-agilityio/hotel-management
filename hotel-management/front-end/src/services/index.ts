export * from "./Rate/rateServices";
export * from "./Room/roomService";
