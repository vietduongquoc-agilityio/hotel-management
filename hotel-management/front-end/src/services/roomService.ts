import axios from "axios";

//InterFace
import { NewRoomData } from "@/Interface/Room";

const BASE_URL = process.env.VITE_BASE_URL;

export const getRooms = async (
  page: number,
  pageSize: number,
  field?: string,
  value?: string
) => {
  try {
    const filters = field && value ? { [`filters[${field}]`]: value } : {};
    const response = await axios.get(`${BASE_URL}/rooms`, {
      params: {
        ...filters,
        "sort[0]": "createdAt:desc",
        "pagination[page]": page,
        "pagination[pageSize]": pageSize,
      },
    });

    if (response.data && response.data.data && response.data.meta.pagination) {
      return {
        rooms: response.data.data,
        pagination: response.data.meta.pagination,
      };
    }
    return response.data;
  } catch (error) {
    console.error("Error fetching room data", error);
    throw error;
  }
};

export const createRoomApi = async (roomData: NewRoomData) => {
  try {
    const response = await axios.post(`${BASE_URL}/rooms`, {
      data: roomData,
    });
    return response.data;
  } catch (error) {
    console.error("Error in createRoom:", error);
    throw error;
  }
};

export const getRoomById = (roomId: string) =>
  axios.get(`${BASE_URL}/rooms/${roomId}`).then((response) => response.data);

export const updateRoom = (roomId: string, roomData: NewRoomData) => {
  if (!roomId) throw new Error("Missing document ID for room update.");
  return axios
    .put(`${BASE_URL}/rooms/${roomId}`, { data: roomData })
    .then((response) => response.data);
};

export const deleteRoom = async (roomId: string) => {
  const response = await axios.delete(`${BASE_URL}/rooms/${roomId}`);
  return response.data;
};
