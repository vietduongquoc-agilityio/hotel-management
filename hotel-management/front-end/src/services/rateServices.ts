/* eslint-disable @typescript-eslint/no-unused-vars */
import axios from "axios";

// InterFace
import { NewRateData } from "@/interfaces/Rate";

const BASE_URL = process.env.VITE_BASE_URL;
// Rate Service
export const getRates = async (page: number, pageSize: number) => {
  try {
    const response = await axios.get(`${BASE_URL}/rates`, {
      params: {
        "sort[0]": "createdAt:desc",
        "pagination[page]": page,
        "pagination[pageSize]": pageSize,
      },
    });
    return response.data;
  } catch (error) {
    console.error("Error fetching rate data", error);
    throw error;
  }
};

// Check the list of rooms using a `rate` tool
export const getRoomsUsingRate = async (rateId: string) => {
  try {
    const response = await axios.get(`${BASE_URL}/rooms`, {
      params: { rateId },
    });
    return response.data;
  } catch (error) {
    console.error("Error fetching rooms using rate", error);
    throw error;
  }
};

export const createRateApi = async (rateData: NewRateData) => {
  try {
    const response = await axios.post(`${BASE_URL}/rates`, {
      data: rateData,
    });
    return response.data;
  } catch (error) {
    console.error("Error in createRate:", error);
    throw error;
  }
};

export const updateRate = (rateId: string, rateData: NewRateData) => {
  if (!rateId) throw new Error("Missing document ID for rate update.");
  return axios
    .put(`${BASE_URL}/rates/${rateId}`, { data: rateData })
    .then((response) => response.data);
};

export const deleteRate = async (rateId: string) => {
  const response = await axios.delete(`${BASE_URL}/rates/${rateId}`);
  return response.data;
};
