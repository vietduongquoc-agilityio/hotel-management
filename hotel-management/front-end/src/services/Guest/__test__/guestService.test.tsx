import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { getGuests, createGuestApi, updateGuest, deleteGuest } from "@/services";
import { NewGuestData } from "@/interfaces";

jest.mock("@chakra-ui/react", () => ({
  createStandaloneToast: jest.fn(() => ({
    toast: jest.fn(),
  })),
}));

const mock = new MockAdapter(axios);
const BASE_URL = process.env.VITE_BASE_URL;

const mockToast = require("@chakra-ui/react").createStandaloneToast();
const { toast } = mockToast;

describe("guestService", () => {
  afterEach(() => {
    mock.reset();
    jest.clearAllMocks();
  });

  describe("getGuests", () => {
    it("should fetch guests with pagination and filters and match snapshot", async () => {
      const mockResponse = {
        data: [
          { id: "1", attributes: { guestName: "John Doe", stay: 3 } },
          { id: "2", attributes: { guestName: "Jane Smith", stay: 2 } },
        ],
        meta: { pagination: { page: 1, pageSize: 10, pageCount: 1, total: 2 } },
      };

      const filters = { roomType: "Deluxe" };
      mock.onGet(`${BASE_URL}/guests`).reply(200, mockResponse);

      const result = await getGuests(1, 10, filters);
      expect(result).toMatchSnapshot();
    });

    it("should handle error when fetching guests", async () => {
      mock.onGet(`${BASE_URL}/guests`).reply(500);

      const result = await getGuests(1, 10);
      expect(result).toEqual({ message: "Error fetching guest data", data: null });
    });
  });

  describe("createGuestApi", () => {
    it("should create a new guest and match snapshot", async () => {
      const guestData: NewGuestData = {
        guestName: "John Doe",
        roomType: "Deluxe",
        stay: 3,
        price: 200,
        registrationNumber: "REG123",
        totalAmount: 600,
        checkInDate: new Date("2024-12-25"),
      };

      const mockResponse = { data: { id: "1", attributes: { ...guestData } } };
      mock.onPost(`${BASE_URL}/guests`).reply(200, mockResponse);

      const result = await createGuestApi(guestData);
      expect(result).toMatchSnapshot();
    });

    it("should handle error when creating a guest", async () => {
      const guestData: NewGuestData = {
        guestName: "John Doe",
        roomType: "Deluxe",
        stay: 3,
        price: 200,
        registrationNumber: "REG123",
        totalAmount: 600,
        checkInDate: new Date("2024-12-25"),
      };

      mock.onPost(`${BASE_URL}/guests`).reply(500);

      const result = await createGuestApi(guestData);
      expect(result).toEqual({ message: "Error updated guest details", data: null });
    });
  });

  describe("updateGuest", () => {
    it("should update a guest by ID and match snapshot", async () => {
      const guestId = "1";
      const guestData: NewGuestData = {
        guestName: "Jane Smith",
        roomType: "Suite",
        stay: 5,
        price: 300,
        registrationNumber: "REG456",
        totalAmount: 1500,
        checkInDate: new Date("2024-12-20"),
      };

      const mockResponse = { data: { id: guestId, attributes: { ...guestData } } };
      mock.onPut(`${BASE_URL}/guests/${guestId}`).reply(200, mockResponse);

      const result = await updateGuest(guestId, guestData);
      expect(result).toMatchSnapshot();
    });

    it("should throw error if guestId is missing", async () => {
      const guestData: NewGuestData = {
        guestName: "Jane Smith",
        roomType: "Suite",
        stay: 5,
        price: 300,
        registrationNumber: "REG456",
        totalAmount: 1500,
        checkInDate: new Date("2024-12-20"),
      };

      await expect(updateGuest("", guestData)).rejects.toThrow(
        "Missing document ID for guest update."
      );
      expect(toast).toHaveBeenCalledWith(
        expect.objectContaining({
          title: "Error",
          description: "Missing document ID for guest update.",
        })
      );
    });

    it("should handle error when updating a guest", async () => {
      const guestId = "1";
      const guestData: NewGuestData = {
        guestName: "Jane Smith",
        roomType: "Suite",
        stay: 5,
        price: 300,
        registrationNumber: "REG456",
        totalAmount: 1500,
        checkInDate: new Date("2024-12-20"),
      };

      mock.onPut(`${BASE_URL}/guests/${guestId}`).reply(500);

      const result = await updateGuest(guestId, guestData);
      expect(result).toEqual({ message: "Error updated guest details", data: null });
    });
  });

  describe("deleteGuest", () => {
    it("should delete a guest by ID and match snapshot", async () => {
      const guestId = "1";
      const mockResponse = { message: "Guest deleted successfully" };

      mock.onDelete(`${BASE_URL}/guests/${guestId}`).reply(200, mockResponse);

      const result = await deleteGuest(guestId);
      expect(result).toMatchSnapshot();
    });

    it("should handle error when deleting a guest", async () => {
      const guestId = "1";

      mock.onDelete(`${BASE_URL}/guests/${guestId}`).reply(500);

      const result = await deleteGuest(guestId);
      expect(result).toEqual({ message: "Error deleted guest details", data: null });
    });
  });
});
