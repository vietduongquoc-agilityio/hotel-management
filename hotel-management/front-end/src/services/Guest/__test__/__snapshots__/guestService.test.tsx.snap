// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`guestService createGuestApi should create a new guest and match snapshot 1`] = `
{
  "data": {
    "attributes": {
      "checkInDate": "2024-12-25T00:00:00.000Z",
      "guestName": "John Doe",
      "price": 200,
      "registrationNumber": "REG123",
      "roomType": "Deluxe",
      "stay": 3,
      "totalAmount": 600,
    },
    "id": "1",
  },
}
`;

exports[`guestService deleteGuest should delete a guest by ID and match snapshot 1`] = `
{
  "data": {
    "message": "Guest deleted successfully",
  },
  "message": "Guest deleted successfully",
}
`;

exports[`guestService getGuests should fetch guests with pagination and filters and match snapshot 1`] = `
{
  "guests": [
    {
      "attributes": {
        "guestName": "John Doe",
        "stay": 3,
      },
      "id": "1",
    },
    {
      "attributes": {
        "guestName": "Jane Smith",
        "stay": 2,
      },
      "id": "2",
    },
  ],
  "pagination": {
    "page": 1,
    "pageCount": 1,
    "pageSize": 10,
    "total": 2,
  },
}
`;

exports[`guestService updateGuest should update a guest by ID and match snapshot 1`] = `
{
  "data": {
    "data": {
      "attributes": {
        "checkInDate": "2024-12-20T00:00:00.000Z",
        "guestName": "Jane Smith",
        "price": 300,
        "registrationNumber": "REG456",
        "roomType": "Suite",
        "stay": 5,
        "totalAmount": 1500,
      },
      "id": "1",
    },
  },
  "message": "Guest updated successfully",
}
`;
